### PR TITLE
TRUNK-4805 | Retrospective stop of orders

### PR DESCRIPTION
**Pull request by aj-jaswanth**
Issue is reopened as a scenario was not considered in the previous commit.

### CODE REVIEW

File: api/src/main/java/org/openmrs/api/impl/OrderServiceImpl.java [hunk 1]
```diff
@@ -337,9 +337,13 @@ private void discontinueExistingOrdersIfNecessary(Order order, Boolean isRetrosp
 			return;
 		}
 
+		Date asOfDate = null;
+		if (isRetrospective)
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/api/impl/OrderServiceImpl.java]: Our convention is to wrap if clauses in curly braces

File: api/src/main/java/org/openmrs/api/impl/OrderServiceImpl.java [hunk 2]
```diff
@@ -337,9 +337,13 @@ private void discontinueExistingOrdersIfNecessary(Order order, Boolean isRetrosp
...
+			asOfDate = order.getDateActivated();
+
 		//Mark first order found corresponding to this DC order as discontinued.
 		List<? extends Order> orders = getActiveOrders(order.getPatient(), order.getOrderType(), order.getCareSetting(),
-		    null);
+		    asOfDate);
```
Comment (wluyima) [inline api/src/main/java/org/openmrs/api/impl/OrderServiceImpl.java]: I think i would disagree with doing this, discontinuing existing  orders is any only makes when doing real time order entry

Comment (vinayvenu) [inline api/src/main/java/org/openmrs/api/impl/OrderServiceImpl.java]: @wluyima If a user wants to record the fact that a discontinue order was issued on a certain day, how would he/she do it?

File: api/src/test/java/org/openmrs/api/OrderServiceTest.java [hunk 3]
```diff
@@ -3109,5 +3109,52 @@ public void saveOrder_shouldRevisePreviousNonCodedOrderIfItIsAlreadyExisting() t
...
+
+		Order discontinueationOrder = originalOrder.cloneForDiscontinuing();
+		discontinueationOrder.setPreviousOrder(originalOrder);
+		discontinueationOrder.setEncounter(encounterService.getEncounter(17));
+		discontinueationOrder.setOrderer(providerService.getProvider(1));
+		discontinueationOrder.setDateActivated(discontinuationOrderDate);
+		orderService.saveRetrospectiveOrder(discontinueationOrder, null);
+
+		assertNotNull(originalOrder.getDateStopped());
+		assertEquals(discontinueationOrder.getAutoExpireDate(), discontinueationOrder.getDateActivated());
```
Comment (wluyima) [inline api/src/test/java/org/openmrs/api/OrderServiceTest.java]: Correct me if am wrong, shouldn't you instead be testing that the originalOrder.dateStopped is equal to discontinuationOrderDate.getDateActivated() but less by a moment?

File: api/src/test/java/org/openmrs/api/OrderServiceTest.java [hunk 4]
```diff
@@ -3109,5 +3109,52 @@ public void saveOrder_shouldRevisePreviousNonCodedOrderIfItIsAlreadyExisting() t
...
+	}
+
+	@Test
+	public void saveRetrospectiveOrder_shouldDiscontinueAndStopActiveOrderInRetrospectiveEntry() throws Exception {
```
Comment (wluyima) [inline api/src/test/java/org/openmrs/api/OrderServiceTest.java]: I see no difference between this test and the one above except the test method name

File: api/src/main/java/org/openmrs/api/impl/OrderServiceImpl.java (hunk 2 above)
Comment (wluyima) [inline api/src/main/java/org/openmrs/api/impl/OrderServiceImpl.java]: On the design call, we agreed to what you're saying so you can ignore my comment, there is more details on the ticket, see Darius' last comment or the updated ticket description under dev notes

File: api/src/test/java/org/openmrs/api/OrderServiceTest.java (hunk 4 above)
Comment (aj-jaswanth) [inline api/src/test/java/org/openmrs/api/OrderServiceTest.java]: First one has a previous order. The second doesn't, so it tries to stop an active order by looking up the active orders present at that time.

Comment (wluyima) [inline api/src/test/java/org/openmrs/api/OrderServiceTest.java]: Ok, thanks!

File: api/src/test/java/org/openmrs/api/OrderServiceTest.java (hunk 3 above)
Comment (wluyima) [inline api/src/test/java/org/openmrs/api/OrderServiceTest.java]: Did you see this comment?

**Comment by dkayiwa** [comment]
@aj-jaswanth can you pull the latest upstream changes and resolve the conflicts?

**Pull request closed without merging**